library/threading.c

```diff
@@ -69,69 +69,115 @@ static int threading_mutex_unlock_pthread( mbedtls_threading_mutex_t *mutex )
+#define MBEDTLS_UNLOCK_FREE_MACRO     threading_mutex_fail
+
+#endif /* !MBEDTLS_MUTEX_INIT_MACRO && !MBEDTLS_MUTEX_FREE_MACRO &&
+        * !MBEDTLS_MUTEX_LOCK_MACRO && !MBEDTLS_UNLOCK_FREE_MACRO */
 
 /*
  * Set functions pointers and initialize global mutexes
  */
-void mbedtls_threading_set_alt( void (*mutex_init)( mbedtls_threading_mutex_t * ),
-                       void (*mutex_free)( mbedtls_threading_mutex_t * ),
-                       int (*mutex_lock)( mbedtls_threading_mutex_t * ),
-                       int (*mutex_unlock)( mbedtls_threading_mutex_t * ) )
+void mbedtls_threading_set_alt( mbedtls_mutex_init_t* mutex_init,
+                                mbedtls_mutex_free_t* mutex_free,
+                                mbedtls_mutex_lock_t* mutex_lock,
+                                mbedtls_mutex_unlock_t* mutex_unlock )
 {
     mbedtls_mutex_init = mutex_init;
     mbedtls_mutex_free = mutex_free;
     mbedtls_mutex_lock = mutex_lock;
     mbedtls_mutex_unlock = mutex_unlock;
 
+#if defined(MBEDTLS_FS_IO)
     mbedtls_mutex_init( &mbedtls_threading_readdir_mutex );
+#endif /* MBEDTLS_FS_IO */
+
+#if defined(MBEDTLS_HAVE_TIME_DATE)
     mbedtls_mutex_init( &mbedtls_threading_gmtime_mutex );
+#endif /* MBEDTLS_HAVE_TIME_DATE */
 }
 
 /*
  * Free global mutexes
  */
 void mbedtls_threading_free_alt( void )
 {
+#if defined(MBEDTLS_FS_IO)
     mbedtls_mutex_free( &mbedtls_threading_readdir_mutex );
+#endif /* MBEDTLS_FS_IO */
+
+#if defined(MBEDTLS_HAVE_TIME_DATE)
     mbedtls_mutex_free( &mbedtls_threading_gmtime_mutex );
+#endif /* MBEDTLS_HAVE_TIME_DATE */
 }
+
+mbedtls_mutex_init_t* mbedtls_mutex_init =
+                            ( mbedtls_mutex_init_t* )MBEDTLS_MUTEX_INIT_MACRO;
+mbedtls_mutex_free_t* mbedtls_mutex_free =
+                            ( mbedtls_mutex_free_t* )MBEDTLS_MUTEX_FREE_MACRO;
+mbedtls_mutex_lock_t* mbedtls_mutex_lock =
+                            ( mbedtls_mutex_lock_t* )MBEDTLS_MUTEX_LOCK_MACRO;
+mbedtls_mutex_unlock_t* mbedtls_mutex_unlock =
+                        ( mbedtls_mutex_unlock_t* )MBEDTLS_MUTEX_UNLOCK_MACRO;
+
 #endif /* MBEDTLS_THREADING_ALT */
 
 /*
  * Define global mutexes
  */
-#ifndef MUTEX_INIT
-#define MUTEX_INIT
-#endif
-mbedtls_threading_mutex_t mbedtls_threading_readdir_mutex MUTEX_INIT;
-mbedtls_threading_mutex_t mbedtls_threading_gmtime_mutex MUTEX_INIT;
+#ifdef MBEDTLS_MUTEX_INITIALIZER
+
+#if defined(MBEDTLS_FS_IO)
+mbedtls_threading_mutex_t mbedtls_threading_readdir_mutex = MBEDTLS_MUTEX_INITIALIZER;
+#endif /* MBEDTLS_FS_IO */
+
+#if defined(MBEDTLS_HAVE_TIME_DATE)
+mbedtls_threading_mutex_t mbedtls_threading_gmtime_mutex = MBEDTLS_MUTEX_INITIALIZER;
+#endif /* MBEDTLS_HAVE_TIME_DATE */
+
+#else
+
+#if defined(MBEDTLS_FS_IO)
+mbedtls_threading_mutex_t mbedtls_threading_readdir_mutex;
+#endif /* MBEDTLS_FS_IO */
+
+#if defined(MBEDTLS_HAVE_TIME_DATE)
+mbedtls_threading_mutex_t mbedtls_threading_gmtime_mutex;
+#endif /* MBEDTLS_HAVE_TIME_DATE */
```

Comment:
If you look at the lines immediately preceding these, you'll see that if `MBEDTLS_MUTEX_INITIALIZER` is not defined than you don't initialise them statically, and if it is defined you do. The choice is down to whomever is providing the alternative implementation of the threading primitives.

`mbedtls_mutex_init()` is called extensively through the library whenever the context a mutex is held in, is initialised.

There are two global mutexes, initialised in `threading.c` statically. If they need to be initialised they'll need to be initialised with `MBEDTLS_MUTEX_INITIALIZER` or outside the library. 

Or we could initialise them in the platform initialisation?

Reply:
Ok, if that is our policy it would be good to document that somewhere visible as you commented below. I looked around in the threading files, config.h and knowledgebase article and could not find anything that warns the user/porter that if there is no static initialiser then the user is expected to initialise manually from outside the library or the platform initialisation code.

Reply:
IMO it's bad to require users to manually initialize global mutexes, because that means they need to change their code whenever we change the set of global mutexes. Currently they're initialized in `mbedtls_threading_set_alt()`. If there are use case where `MBEDTLS_MUTEX_INITIALIZER` is not defined and this function is not called, then I suggest we split its last two lines into a new public function and require users in that situation to call that function, instead of directly accessing global variables.

Reply:
It looks like the main incentive to use the macros is to avoid calling `mbedtls_threading_set_alt()`. So we might as well require an all or nothing approach. (We could enforce in check-config.h that either all of these macros are defined or none of them)